modules/store/src/provide_store.ts

```diff
@@ -0,0 +1,333 @@
+import {
+  Inject,
+  InjectionToken,
+  Injector,
+  Optional,
+  SkipSelf,
+  ImportedNgModuleProviders,
+  ENVIRONMENT_INITIALIZER,
+  inject,
+  InjectFlags,
+  Provider,
+} from '@angular/core';
+import {
+  Action,
+  ActionReducer,
+  ActionReducerMap,
+  StoreFeature,
+} from './models';
+import { combineReducers, createReducerFactory } from './utils';
+import {
+  INITIAL_STATE,
+  INITIAL_REDUCERS,
+  _INITIAL_REDUCERS,
+  REDUCER_FACTORY,
+  _REDUCER_FACTORY,
+  STORE_FEATURES,
+  _INITIAL_STATE,
+  META_REDUCERS,
+  _STORE_REDUCERS,
+  FEATURE_REDUCERS,
+  _FEATURE_REDUCERS,
+  _FEATURE_REDUCERS_TOKEN,
+  _STORE_FEATURES,
+  _FEATURE_CONFIGS,
+  USER_PROVIDED_META_REDUCERS,
+  _RESOLVED_META_REDUCERS,
+  _ROOT_STORE_GUARD,
+  _ACTION_TYPE_UNIQUENESS_CHECK,
+} from './tokens';
+import { ACTIONS_SUBJECT_PROVIDERS, ActionsSubject } from './actions_subject';
+import {
+  REDUCER_MANAGER_PROVIDERS,
+  ReducerManager,
+  ReducerObservable,
+} from './reducer_manager';
+import {
+  SCANNED_ACTIONS_SUBJECT_PROVIDERS,
+  ScannedActionsSubject,
+} from './scanned_actions_subject';
+import { STATE_PROVIDERS } from './state';
+import { STORE_PROVIDERS, Store } from './store';
+import {
+  provideRuntimeChecks,
+  checkForActionTypeUniqueness,
+} from './runtime_checks';
+import {
+  FeatureSlice,
+  RootStoreConfig,
+  StoreConfig,
+  _concatMetaReducers,
+  _createFeatureReducers,
+  _createFeatureStore,
+  _createStoreReducers,
+  _initialStateFactory,
+  _provideForRootGuard,
+} from './store_config';
+
+/**
+ * Environment Initializer used in the feature
+ * providers to register state features
+ */
+const ENVIRONMENT_STATE_PROVIDER: Provider[] = [
+  {
+    provide: ENVIRONMENT_INITIALIZER,
+    multi: true,
+    deps: [],
+    useFactory() {
+      return () => {
+        const features = inject<StoreFeature<any, any>[]>(_STORE_FEATURES);
+        const featureReducers =
+          inject<ActionReducerMap<any>[]>(FEATURE_REDUCERS);
+        const reducerManager = inject(ReducerManager);
+        inject(_ACTION_TYPE_UNIQUENESS_CHECK, InjectFlags.Optional);
+
+        const feats = features.map((feature, index) => {
+          const featureReducerCollection = featureReducers.shift();
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          const reducers = featureReducerCollection! /*TODO(#823)*/[index];
+
+          return {
+            ...feature,
+            reducers,
+            initialState: _initialStateFactory(feature.initialState),
+          };
+        });
+
+        reducerManager.addFeatures(feats);
+      };
+    },
+  },
+];
```

Comment:
We could move feature registration logic to a separate service. This service can be used in `provideEffects` to ensure that the store feature is registered before the effects.

```suggestion
@Injectable()
export class RegisterStateService {
  constructor() {
    // tokens can be injected via constructor 👇
    const features = inject<StoreFeature<any, any>[]>(_STORE_FEATURES);
    const featureReducers =
      inject<ActionReducerMap<any>[]>(FEATURE_REDUCERS);
    const reducerManager = inject(ReducerManager);
    inject(_ACTION_TYPE_UNIQUENESS_CHECK, InjectFlags.Optional);

     // feature registration logic can be moved to a separate method 👇
    const feats = features.map((feature, index) => {
      const featureReducerCollection = featureReducers.shift();
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const reducers = featureReducerCollection! /*TODO(#823)*/[index];

      return {
        ...feature,
        reducers,
        initialState: _initialStateFactory(feature.initialState),
      };
    });

    reducerManager.addFeatures(feats);
  }
}

/**
 * Environment Initializer used in the feature
 * providers to register state features
 */
const ENVIRONMENT_STATE_PROVIDER: Provider[] = [
  RegisterStateService,
  {
    provide: ENVIRONMENT_INITIALIZER,
    multi: true,
    useFactory() {
      return () => inject(RegisterStateService);
    },
  },
];
```

Then, we can inject `RegisterStateService` in `provideEffects` to ensure that the store feature is registered before the effects:

```ts
export function provideEffects(effects: Type<unknown>[]): Provider[] {
  return [
    effects,
    {
      provide: ENVIRONMENT_INITIALIZER,
      multi: true,
      useValue: () => {
        inject(RegisterStateService, InjectFlags.Optional);

        // register effects
      },
    },
  ];
}
```

Reply:
Agreed. What about using another InjectionToken with a factory instead of a service?

Reply:
Sounds good 👍